File: cxx/src/api/game.cpp

#include "api/game.h"

#include "api/convert.h"
#include "api/exception.h"

namespace {
loot::GameType convert(loot::rust::GameType gameType) {
  switch (gameType) {
    case loot::rust::GameType::tes3:
      return loot::GameType::tes3;
    case loot::rust::GameType::tes4:
      return loot::GameType::tes4;
    case loot::rust::GameType::tes5:
      return loot::GameType::tes5;
    case loot::rust::GameType::tes5se:
      return loot::GameType::tes5se;
    case loot::rust::GameType::tes5vr:
      return loot::GameType::tes5vr;
    case loot::rust::GameType::fo3:
      return loot::GameType::fo3;
    case loot::rust::GameType::fonv:
      return loot::GameType::fonv;
    case loot::rust::GameType::fo4:
      return loot::GameType::fo4;
    case loot::rust::GameType::fo4vr:
      return loot::GameType::fo4vr;
    case loot::rust::GameType::starfield:
      return loot::GameType::starfield;
    case loot::rust::GameType::openmw:
      return loot::GameType::openmw;
    default:
      throw std::logic_error("Unsupported GameType value");
  }
}

loot::rust::GameType convert(loot::GameType gameType) {
  switch (gameType) {
    case loot::GameType::tes3:
      return loot::rust::GameType::tes3;
    case loot::GameType::tes4:
      return loot::rust::GameType::tes4;
    case loot::GameType::tes5:
      return loot::rust::GameType::tes5;
    case loot::GameType::tes5se:
      return loot::rust::GameType::tes5se;
    case loot::GameType::tes5vr:
      return loot::rust::GameType::tes5vr;
    case loot::GameType::fo3:
      return loot::rust::GameType::fo3;
    case loot::GameType::fonv:
      return loot::rust::GameType::fonv;
    case loot::GameType::fo4:
      return loot::rust::GameType::fo4;
    case loot::GameType::fo4vr:
      return loot::rust::GameType::fo4vr;
    case loot::GameType::starfield:
      return loot::rust::GameType::starfield;
    case loot::GameType::openmw:
      return loot::rust::GameType::openmw;
    default:
      throw std::logic_error("Unsupported GameType value");
  }
}

rust::Box<loot::rust::Game> constructGame(
    const loot::GameType gameType,
    const std::filesystem::path& gamePath,
    const std::filesystem::path& localDataPath) {
  try {
    if (localDataPath.empty()) {
      return std::move(
          loot::rust::new_game(convert(gameType), gamePath.u8string()));
    } else {
      return std::move(loot::rust::new_game_with_local_path(
          convert(gameType), gamePath.u8string(), localDataPath.u8string()));
    }
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(loot::mapError(e));
  }
}

std::vector<::rust::Str> as_str_refs(const std::vector<std::string>& vector) {
  std::vector<::rust::Str> strings;
  for (const auto& str : vector) {
    strings.push_back(str);
  }

  return strings;
}
}

namespace loot {
Game::Game(const GameType gameType,
           const std::filesystem::path& gamePath,
           const std::filesystem::path& localDataPath) :
    game_(constructGame(gameType, gamePath, localDataPath)),
    database_(game_->database()) {}

GameType Game::GetType() const {
  try {
    return ::convert(game_->game_type());
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

const DatabaseInterface& Game::GetDatabase() const { return database_; }

DatabaseInterface& Game::GetDatabase() { return database_; }

std::vector<std::filesystem::path> Game::GetAdditionalDataPaths() const {
  try {
    std::vector<std::filesystem::path> paths;
    for (const auto& path_str : game_->additional_data_paths()) {
      paths.push_back(
          std::filesystem::u8path(path_str.begin(), path_str.end()));
    }

    return paths;
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

void Game::SetAdditionalDataPaths(
    const std::vector<std::filesystem::path>& additionalDataPaths) {
  std::vector<::rust::String> path_strings;
  std::vector<::rust::Str> path_strs;
  for (const auto& path : additionalDataPaths) {
    path_strings.push_back(path.u8string());
    path_strs.push_back(path_strings.back());
  }
  try {
    game_->set_additional_data_paths(
        ::rust::Slice<const ::rust::Str>(path_strs));
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

bool Game::IsValidPlugin(const std::filesystem::path& pluginPath) const {
  return game_->is_valid_plugin(pluginPath.u8string());
}

void Game::LoadPlugins(const std::vector<std::filesystem::path>& pluginPaths,
                       bool loadHeadersOnly) {
  std::vector<::rust::String> path_strings;
  std::vector<::rust::Str> path_strs;
  for (const auto& path : pluginPaths) {
    path_strings.push_back(path.u8string());
    path_strs.push_back(path_strings.back());
  }

  try {
    if (loadHeadersOnly) {
      game_->load_plugin_headers(::rust::Slice<const ::rust::Str>(path_strs));
    } else {
      game_->load_plugins(::rust::Slice<const ::rust::Str>(path_strs));
    }
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }

  for (const auto& path : pluginPaths) {
    auto key = Filename(path.filename().u8string());
    auto it = plugins_.find(key);
    if (it != plugins_.end()) {
      plugins_.erase(it);
    }
  }
}

void Game::ClearLoadedPlugins() {
  game_->clear_loaded_plugins();
  plugins_.clear();
}

const PluginInterface* Game::GetPlugin(const std::string& pluginName) const {
  const auto pluginOpt = game_->plugin(pluginName);
  if (!pluginOpt->is_some()) {
    return nullptr;
  }

  auto key = Filename(pluginName);
  const auto it = plugins_.find(key);
  if (it != plugins_.end()) {
    return it->second.get();
  }

  try {
    auto plugin = std::make_shared<Plugin>(std::move(pluginOpt->as_ref()));
    const auto result = plugins_.emplace(key, plugin);

    return result.first->second.get();
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

std::vector<const PluginInterface*> Game::GetLoadedPlugins() const {
  std::vector<const PluginInterface*> plugins;
  for (const auto& pluginRef : game_->loaded_plugins()) {
    const auto plugin =
        std::make_shared<Plugin>(std::move(pluginRef.boxed_clone()));

    const auto result =
        plugins_.insert_or_assign(Filename(plugin->GetName()), plugin);
    plugins.push_back(result.first->second.get());
  }

  return plugins;
}

std::vector<std::string> Game::SortPlugins(
    const std::vector<std::string>& pluginFilenames) {
  const auto strs = as_str_refs(pluginFilenames);

  try {
    const auto results = game_->sort_plugins(::rust::Slice(strs));

    return convert<std::string>(results);
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

void Game::LoadCurrentLoadOrderState() {
  try {
    game_->load_current_load_order_state();
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

bool Game::IsLoadOrderAmbiguous() const {
  try {
    return game_->is_load_order_ambiguous();
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

std::filesystem::path Game::GetActivePluginsFilePath() const {
  try {
    const auto path_string = game_->active_plugins_file_path();
    return std::filesystem::u8path(path_string.begin(), path_string.end());
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}

bool Game::IsPluginActive(const std::string& pluginName) const {
  return game_->is_plugin_active(pluginName);
}

std::vector<std::string> Game::GetLoadOrder() const {
  return convert<std::string>(game_->load_order());
}

void Game::SetLoadOrder(const std::vector<std::string>& loadOrder) {
  const auto strs = as_str_refs(loadOrder);

  try {
    game_->set_load_order(::rust::Slice(strs));
  } catch (const ::rust::Error& e) {
    std::rethrow_exception(mapError(e));
  }
}
}

File: cxx/src/game.rs
use std::path::Path;

use delegate::delegate;
use libloot_ffi_errors::UnsupportedEnumValueError;

use crate::{Plugin, VerboseError, database::Database, ffi::GameType, plugin::OptionalPlugin};

impl TryFrom<libloot::GameType> for GameType {
    type Error = UnsupportedEnumValueError;

    fn try_from(value: libloot::GameType) -> Result<Self, Self::Error> {
        match value {
            libloot::GameType::TES4 => Ok(GameType::tes4),
            libloot::GameType::TES5 => Ok(GameType::tes5),
            libloot::GameType::FO3 => Ok(GameType::fo3),
            libloot::GameType::FONV => Ok(GameType::fonv),
            libloot::GameType::FO4 => Ok(GameType::fo4),
            libloot::GameType::TES5SE => Ok(GameType::tes5se),
            libloot::GameType::FO4VR => Ok(GameType::fo4vr),
            libloot::GameType::TES5VR => Ok(GameType::tes5vr),
            libloot::GameType::TES3 => Ok(GameType::tes3),
            libloot::GameType::Starfield => Ok(GameType::starfield),
            libloot::GameType::OpenMW => Ok(GameType::openmw),
            _ => Err(UnsupportedEnumValueError),
        }
    }
}

impl TryFrom<GameType> for libloot::GameType {
    type Error = UnsupportedEnumValueError;

    fn try_from(value: GameType) -> Result<Self, Self::Error> {
        match value {
            GameType::tes4 => Ok(libloot::GameType::TES4),
            GameType::tes5 => Ok(libloot::GameType::TES5),
            GameType::fo3 => Ok(libloot::GameType::FO3),
            GameType::fonv => Ok(libloot::GameType::FONV),
            GameType::fo4 => Ok(libloot::GameType::FO4),
            GameType::tes5se => Ok(libloot::GameType::TES5SE),
            GameType::fo4vr => Ok(libloot::GameType::FO4VR),
            GameType::tes5vr => Ok(libloot::GameType::TES5VR),
            GameType::tes3 => Ok(libloot::GameType::TES3),
            GameType::starfield => Ok(libloot::GameType::Starfield),
            GameType::openmw => Ok(libloot::GameType::OpenMW),
            _ => Err(UnsupportedEnumValueError),
        }
    }
}

impl From<Game> for libloot::Game {
    fn from(value: Game) -> Self {
        value.0
    }
}

impl From<libloot::Game> for Game {
    fn from(value: libloot::Game) -> Self {
        Game(value)
    }
}

#[derive(Debug)]
pub struct NotValidUtf8;

impl std::fmt::Display for NotValidUtf8 {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Value was not valid UTF-8")
    }
}

impl std::error::Error for NotValidUtf8 {}

#[derive(Debug)]
#[repr(transparent)]
pub struct Game(libloot::Game);

// CXX doesn't support &Path so use &str instead.
pub fn new_game(game_type: GameType, game_path: &str) -> Result<Box<Game>, VerboseError> {
    libloot::Game::new(game_type.try_into()?, Path::new(game_path))
        .map(|g| Box::new(g.into()))
        .map_err(Into::into)
}

pub fn new_game_with_local_path(
    game_type: GameType,
    game_path: &str,
    game_local_path: &str,
) -> Result<Box<Game>, VerboseError> {
    libloot::Game::with_local_path(
        game_type.try_into()?,
        Path::new(game_path),
        Path::new(game_local_path),
    )
    .map(|g| Box::new(g.into()))
    .map_err(Into::into)
}

fn path_to_string(path: &Path) -> Result<String, VerboseError> {
    path.to_str()
        .map(|s| s.to_owned())
        .ok_or(NotValidUtf8)
        .map_err(Into::into)
}

fn strings_to_paths<'a>(paths: &'a [&'a str]) -> Vec<&'a Path> {
    paths.iter().map(Path::new).collect()
}

impl Game {
    pub fn game_type(&self) -> Result<GameType, VerboseError> {
        self.0.game_type().try_into().map_err(Into::into)
    }

    pub fn additional_data_paths(&self) -> Result<Vec<String>, VerboseError> {
        self.0
            .additional_data_paths()
            .iter()
            .map(|p| path_to_string(p))
            .collect()
    }

    pub fn set_additional_data_paths(
        &mut self,
        additional_data_paths: &[&str],
    ) -> Result<(), VerboseError> {
        self.0
            .set_additional_data_paths(&strings_to_paths(additional_data_paths))
            .map_err(Into::into)
    }

    pub fn database(&self) -> Box<Database> {
        Box::new(Database::new(self.0.database()))
    }

    pub fn is_valid_plugin(&self, plugin_path: &str) -> bool {
        self.0.is_valid_plugin(Path::new(plugin_path))
    }

    pub fn load_plugins(&mut self, plugin_paths: &[&str]) -> Result<(), VerboseError> {
        self.0
            .load_plugins(&strings_to_paths(plugin_paths))
            .map_err(Into::into)
    }

    pub fn load_plugin_headers(&mut self, plugin_paths: &[&str]) -> Result<(), VerboseError> {
        self.0
            .load_plugin_headers(&strings_to_paths(plugin_paths))
            .map_err(Into::into)
    }

    pub fn plugin(&self, plugin_name: &str) -> Box<OptionalPlugin> {
        Box::new(self.0.plugin(plugin_name).into())
    }

    pub fn loaded_plugins(&self) -> Vec<Plugin> {
        self.0
            .loaded_plugins()
            .into_iter()
            .map(Into::into)
            .collect()
    }

    pub fn sort_plugins(&self, plugin_names: &[&str]) -> Result<Vec<String>, VerboseError> {
        self.0.sort_plugins(plugin_names).map_err(Into::into)
    }

    pub fn load_current_load_order_state(&mut self) -> Result<(), VerboseError> {
        self.0.load_current_load_order_state().map_err(Into::into)
    }

    pub fn is_load_order_ambiguous(&self) -> Result<bool, VerboseError> {
        self.0.is_load_order_ambiguous().map_err(Into::into)
    }

    pub fn active_plugins_file_path(&self) -> Result<String, VerboseError> {
        path_to_string(self.0.active_plugins_file_path())
    }

    pub fn load_order(&self) -> Vec<String> {
        self.0.load_order().iter().map(|s| s.to_string()).collect()
    }

    pub fn set_load_order(&mut self, load_order: &[&str]) -> Result<(), VerboseError> {
        self.0.set_load_order(load_order).map_err(Into::into)
    }

    delegate! {
        to self.0 {
            pub fn clear_loaded_plugins(&mut self);

            pub fn is_plugin_active(&self, plugin_name: &str) -> bool;
        }
    }
}
